svsim/src/main/scala/Workspace.scala

```diff
@@ -317,12 +319,43 @@ final class Workspace(
       commonSettings = commonSettings,
       backendSpecificSettings = backendSpecificSettings
     )
-    val sourceFiles = Seq(primarySourcesPath, generatedSourcesPath)
-      .flatMap(p => Files.walk(Paths.get(p)).iterator().asScala.toSeq)
-      .map(_.toFile)
-      .filter(_.isFile)
-      .filter(commonSettings.fileFilter.orElse { case _ => true })
-      .map { file => workingDirectory.toPath().relativize(file.toPath()).toString() }
+
+    // Find all the source files by walking the build directory.  The behavior
+    // of which directories are visited and which files are included is
+    // controlled by fields in `CommonCompilationSettings`.
+    val sourceFiles = scala.collection.mutable.ArrayBuffer.empty[String]
+    class DirectoryVisitor extends FileVisitor[Path] {
+
+      override def visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
+        val filter: PartialFunction[File, Boolean] = commonSettings.fileFilter.orElse { case _ => true }
+
+        if (filter(file.toFile))
+          sourceFiles += workingDirectory.toPath().relativize(file).toString()
+        FileVisitResult.CONTINUE
+      }
+
+      override def preVisitDirectory(dir: Path, attrs: BasicFileAttributes): FileVisitResult = {
+        val filter: PartialFunction[File, Boolean] = commonSettings.directoryFilter.orElse { case _: File => true }
+
+        if (filter(dir.toFile))
```

Comment:
```suggestion
    val sourceFiles = scala.collection.mutable.ArrayBuffer.empty[String]
    val fileFilter: PartialFunction[File, Boolean] = commonSettings.fileFilter.orElse { case _ => true }
    val directoryFilter: PartialFunction[File, Boolean] = commonSettings.directoryFilter.orElse { case _: File => true }
    class DirectoryVisitor extends FileVisitor[Path] {

      override def visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
        if (fileFilter(file.toFile))
          sourceFiles += workingDirectory.toPath().relativize(file).toString()
        FileVisitResult.CONTINUE
      }

      override def preVisitDirectory(dir: Path, attrs: BasicFileAttributes): FileVisitResult = {
        if (directoryFilter(dir.toFile))
```
This a nit/micro-optimization, but creating the filters is at least 2 object allocations each that could be memoized across all files and directories.

Reply:
Good suggestion. Applied in https://github.com/chipsalliance/chisel/pull/4709/commits/00899a2e647ce7e2af786c4bcf6d2f69a426c568.